unified-runtime/test/adapters/level_zero/enqueue_alloc.cpp

```diff
@@ -0,0 +1,685 @@
+  ASSERT_SUCCESS(urKernelSetArgValue(kernel, 1, sizeof(DATA), nullptr, &DATA));
+  Launch1DRange(ARRAY_SIZE);
+
+  ValidateEnqueueFree(ptr2);
+}
+
+UUR_DEVICE_TEST_SUITE_WITH_PARAM(
+    urL0EnqueueAllocMultiQueueSameDeviceTest,
+    ::testing::ValuesIn({
+        EnqueueAllocMultiQueueTestParam{1024, 256, 8, urEnqueueUSMHostAllocExp,
+                                        uur::GetDeviceUSMHostSupport},
+        EnqueueAllocMultiQueueTestParam{1024, 256, 8,
+                                        urEnqueueUSMSharedAllocExp,
+                                        uur::GetDeviceUSMSingleSharedSupport},
+        EnqueueAllocMultiQueueTestParam{1024, 256, 8,
+                                        urEnqueueUSMDeviceAllocExp,
+                                        uur::GetDeviceUSMDeviceSupport},
+    }),
+    uur::deviceTestWithParamPrinter<EnqueueAllocMultiQueueTestParam>);
+
+TEST_P(urL0EnqueueAllocMultiQueueSameDeviceTest, SuccessMt) {
+  const size_t allocSize = std::get<1>(this->GetParam()).allocSize;
+  const size_t numQueues = std::get<1>(this->GetParam()).numQueues;
+  const size_t iterations = std::get<1>(this->GetParam()).iterations;
+  const auto enqueueUSMAllocFunc =
+      std::get<1>(this->GetParam()).funcParams.enqueueUSMAllocFunc;
+  const auto checkUSMSupportFunc =
+      std::get<1>(this->GetParam()).funcParams.checkUSMSupportFunc;
+
+  ur_device_usm_access_capability_flags_t USMSupport = 0;
+  ASSERT_SUCCESS(checkUSMSupportFunc(device, USMSupport));
+  if (!(USMSupport & UR_DEVICE_USM_ACCESS_CAPABILITY_FLAG_ACCESS)) {
+    GTEST_SKIP() << "Selected USM type is not supported.";
+  }
+
+  std::mutex mtx;
+  std::condition_variable cv1;
+  size_t cvCounter1 = numQueues;
+  std::condition_variable cv2;
+  size_t cvCounter2 = numQueues;
+
+  std::vector<std::thread> threads;
+  for (size_t idx = 0; idx < numQueues; idx++) {
+    // Create a thread per queue.
+    threads.emplace_back([&, idx, iterations] {
+      std::vector<void *> ptrs(iterations);
+      for (size_t i = 0; i < iterations; i++) {
+        ASSERT_SUCCESS(enqueueUSMAllocFunc(queues[idx], nullptr, allocSize,
+                                           nullptr, 0, nullptr, &ptrs[i],
+                                           nullptr));
+        ASSERT_NE(ptrs[i], nullptr);
+      }
+
+      ASSERT_SUCCESS(urQueueFinish(queues[idx]));
+
+      // Synchronize threads.
+      {
+        std::unique_lock<std::mutex> lock(mtx);
+        cvCounter1--;
+        if (cvCounter1 > 0) {
+          cv1.wait(lock, [&] { return cvCounter1 == 0; });
+        } else {
+          cv1.notify_all();
+        }
+      }
+
+      uint8_t fillPattern = 0xAF;
+      for (size_t i = 0; i < iterations; i++) {
+        ASSERT_SUCCESS(urEnqueueUSMFill(queues[idx], ptrs[i], 1, &fillPattern,
+                                        allocSize, 0, nullptr, nullptr));
+      }
+
+      ASSERT_SUCCESS(urQueueFinish(queues[idx]));
+
+      // Synchronize threads.
```

Comment:
why do we need to sync threads?

Reply:
I was tinkering with different scenarios. Actually removing those synchronization points test the execution paths better.
Done.